client/rpc-spec-v2/src/chain_head/subscription.rs

```diff
@@ -94,13 +104,17 @@ impl<Block: BlockT> SubscriptionHandle<Block> {
 
 		if inner.blocks.len() == inner.max_pinned_blocks {
 			// We have reached the limit. However, the block can be already inserted.
+
 			if inner.blocks.contains(&hash) {
```

Comment:
Here is a logic error? If we try to pin the same block twice, the first `unpin` will unpin it in the backend (assuming it is only pinned by this subscription). Is this documented in the spec that when you receive the same block as best and some seconds later as finalized (without having called unpin in between), you only need to call unpin once?

Reply:
Nice catch! I couldn't find something super specific in the RPC spec regarding this.

From the [spec](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html):
> Note: A JSON-RPC client should call chainHead_unstable_unpin only after it is sure to no longer be interested in a certain block. This typically happens after the block has been finalized or pruned. There is no requirement to call chainHead_unstable_unpin as quickly as possible.

I was also thinking along the lines: the [newBlock](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#newblock) event is where the block is declared and pinned; and [bestBlockChanged](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#bestblockchanged) and [finalized](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#finalized) are updates about what happens with that block.

I think we could update the spec documentation for the `unpin` method to state that only block hashes reported by the `newBlock` event are supposed to be unpinned.

What do you think about it?

Reply:
> I was also thinking along the lines: the [newBlock](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#newblock) event is where the block is declared and pinned; and [bestBlockChanged](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#bestblockchanged) and [finalized](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#finalized) are updates about what happens with that block.

It sounds reasonable to me that:
- a block is pinned just prior to a client being given a `newBlock` event, 
- later we might see `best`/`finalised` events emitted re that block (and no pinning happens at this point), 
- at any point after the `newBlock`, whenever a user would like to unpin that block hash they will call "unpin" once against the hash to unpin it.

But, if a `newBlock` is unpinned quickly and _then_ a best/finalized event is emitted for it.. should it need unpinning again? I think offhand I'd go with "it's been unpinned so it stays unpinned, and those events are just notifications about the block you've already declared you don't care about any more".

But that would make it annoying for people to ignore all but best/finalized blocks! Because the simplest logic to do so would be to call "unpin" on all "newBlock" events and not on any best/finalized block events.

"unpin" could potentially take both a block hash _and_ a type (new/best/finalized) so that you can opt-out of all block events you don't care about but still keep pinned the ones you do?

Or, a best/finalized notificiation could pin the block again, and "unpin" just takes the hash, but that runs the risk of race conditions (I "unpin" a newBlock but by the time the unpin request hits the node ,a bestBlock for that hash is emitted).

So two options then are:
- Blocks are _only_ pinned at "newBlock". Unpin will unpin, and best/finalized events may exist for blocks that have already been unpinned.
- new/best/finalized blocks are treated as independent. "unpin" takes a hash and type to unpin exactly one of those.

Reply:
> I was also thinking along the lines: the [newBlock](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#newblock) event is where the block is declared and pinned; and [bestBlockChanged](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#bestblockchanged) and [finalized](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#finalized) are updates about what happens with that block.

I think we should pin on the first event we see (a client can only see bestBlock/finalized without newblock by having connected after the block was imported) and then unpin should be called when they are not interested anymore. We should update the spec to mention this behavior.

Reply:
Yep, we could update the spec to reflect something along the lines:
"Any block hash reported by chainHead's events is guaranteed to be pinned in memory and it requires a single `chainHead_unstable_unpin` call to unpin the block."

The spec states currently:
> The current finalized block reported in the initialized event, and each subsequent block reported with a newBlock event, is automatically considered by the JSON-RPC server as pinned. 

---
> I think we should pin on the first event we see

Yep, that's along the lines of how we pin internally for this `chainhead` subscription.

> a client can only see bestBlock/finalized without newblock by having connected after the block was imported

Before using the subscriptions (best block and finalized) we are:
- generating finalized event (and pin the finalized block)
- generate `newBlock` event for each children of the finalized recursively (and pin the block)

I might be wrong, but I think its entirely possible that we can miss blocks (warp sync etc) and such we always `pin` the block on the first event that we get from the subscriptions. 
In case of errors or jumps, the chainHead should produce the `Stop` events. 

Let me know if that is correct, I ll take an action to create a PR to clarify a bit the spec, thanks!

Reply:
Another way would be to allow users to call `unpin` only for hashes reported by the `Finalized` event: https://github.com/paritytech/json-rpc-interface-spec/issues/32. Let me know if this makes a bit more sense, thanks!

Reply:
> and then unpin should be called when they are not interested anymore.

What if they call `unpin` after receiving eg a `newBlock`, and then the server sends a `finalized` with the same hash? I assume that the block is still unpinned now.

If I was only interested in finalized blocks, and did not want to have any non-finalized blocks pinned any longer than necessary, what would I do?

I gather from chatting with Alex that that what I'd do is hold off on calling "unpin" on anything until I see a `finalized` notification (which also lists all blocks that are "pruned"), and then I could use that list to `unpin` all blocks that are listed as "pruned" now that I know they aren't important any more. (This relies on that list of pruned blocks really just being a list of all blocks not in the finalized chain, since if they are pinned they wouldn't be pruned). Does that make sense?